### ISSUE TITLE
Lambda and method type argument inference with non-denotable type

### ISSUE THREAD

**Issue by maggu2810** [typetools/checker-framework#1464]
After bumping the checker framework from 2.1.10 to 2.1.14 I realized a change behavior.
I don't know if this is an error, but I don't understand the error at all.

Can you please have a look at?

The annotation KeyForButtom states: "Programmers should rarely write this type."

The error report looks like:

```text
src/Test.java:18: error: [[all, keyfor]:lambda.param.type.incompatible] incompatible parameter types in lambda expression.
        addChangedListener(tmp, variable -> System.out.println(variable));
                                ^
  found   : @UnknownKeyFor Variable<capture#902 of ?[ extends @UnknownKeyFor Object super @UnknownKeyFor Void]>
  required: @UnknownKeyFor Variable<capture#902 of ?[ extends @UnknownKeyFor Object super @KeyForBottom Void]>
```

The project can be found here: https://github.com/maggu2810/checker-framework-demo/tree/master/cf01
So the directory `cf01` of that repository.
You can execute `./compile` to reproduce it easily.

Here is the source code of the file, too:

```Java

public class Test {

    public interface Variable<T> {

        void addChangedListener(VariableChangedListener<T> listener);
    }

    public interface VariableChangedListener<T> {

        void variableChanged(final Variable<T> variable);
    }

    protected <T> void addChangedListener(final Variable<T> variable, final VariableChangedListener<T> listener) {
    }

    public void main(final Variable<?> tmp) {
        addChangedListener(tmp, variable -> System.out.println(variable));
    }

}
```

**Comment by maggu2810** [comment]
Anyone that could help?

**Comment by wmdietl** [comment]
Hi Markus,

thanks for reporting this issue! Between 2.1.10 and 2.1.14 we fixed many issues with Java 8 type argument inference and lambdas, but here we have a regression.

The issue is that the lambda parameter `variable` is not correctly inferred in the enhanced target typing location of the polymorphic method.
The inferred type has an incompatible lower bound compared to the type inferred for the invocation of `addChangedListener`.
So I tried making either the lambda parameter or the method invocation type argument explicit, as in:

```java
addChangedListener(tmp, (Variable<?> variable) -> System.out.println(variable));
this.<?>addChangedListener(tmp, variable -> System.out.println(variable));
```

I also tried extracting the lambda expression into a separate variable.

However, none of these work, because `?` is not the right type argument - it would need to be the capture of the wildcard in tmp, but that type isn't denotable.
(The `this.<?>` syntax isn't even well-formed.)

So the only solution I see at the moment is to suppress the warning.
Reference to #979 which will implement proper inference. Keeping this issue open as the example seems quite tricky.

@smillst Do you think we can improve the lower bounds we infer for the lambda parameter and method arguments, to prevent this error? Maybe we are simply inconsistent in the inference of these two cases?
Or do you see a completely different issue?

**Comment by smillst** [comment]
This is definitely a bug that should be fixed with Java 8 type argument inference.

**Comment by smillst** [comment]
We agreed to figure out how to avoid the false positive before #979 is fixed.

**Comment by maggu2810** [comment]
@smillst Sorry if that question is a little bit off-topic, but is there a snapshot repository I can use to check the most recent checker framework artifacts / version?

**Comment by wmdietl** [comment]
We used to have a Jenkins build server that provided build snapshots. To get better github integration we have now transitioned to Travis.ci. As far as I can tell Travis doesn't provide access to the build artifacts.
Do you have a suggestion for a service that would make this easy to do?

**Comment by maggu2810** [comment]
First you need a repository that is able to handle snapshot artifacts.

Couldn't you use Bintray, JCenter and oss.jfrog.org?
Have a look at here:
https://www.jfrog.com/confluence/display/RTF/Deploying+Snapshots+to+oss.jfrog.org#DeployingSnapshotstooss.jfrog.org-GettingStartedwithOJO
If you add your releases not only to Maven Central but also to Bintray and integrate it into JCenter, you can also publish your snapshots to jfrog - for free.

After that I assume you could modify your Travis CI settings to deploy the artifacts after a successful build. Perhaps you could be inspired by https://blog.travis-ci.com/2017-03-30-deploy-maven-travis-ci-packagecloud/
I didn't tried it myself but I don't see any reason why it shouldn't work.
